applicationset/generators/cluster.go

```diff
@@ -105,72 +101,84 @@ func (g *ClusterGenerator) GenerateParams(appSetGenerator *argoappsetv1alpha1.Ap
 				return nil, fmt.Errorf("error appending templated values for local cluster: %w", err)
 			}
 
-			if isFlatMode {
-				clustersParams = append(clustersParams, params)
-			} else {
-				res = append(res, params)
-			}
-
+			paramHolder.append(params)
 			logCtx.WithField("cluster", "local cluster").Info("matched local cluster")
 		}
 	}
 
 	// For each matching cluster secret (non-local clusters only)
 	for _, cluster := range secretsFound {
-		params := map[string]any{}
-
-		params["name"] = string(cluster.Data["name"])
-		params["nameNormalized"] = utils.SanitizeName(string(cluster.Data["name"]))
-		params["server"] = string(cluster.Data["server"])
+		params := g.getClusterParameters(cluster, appSet)
 
-		project, ok := cluster.Data["project"]
-		if ok {
-			params["project"] = string(project)
-		} else {
-			params["project"] = ""
+		err = appendTemplatedValues(appSetGenerator.Clusters.Values, params, appSet.Spec.GoTemplate, appSet.Spec.GoTemplateOptions)
+		if err != nil {
+			return nil, fmt.Errorf("error appending templated values for cluster: %w", err)
 		}
 
-		if appSet.Spec.GoTemplate {
-			meta := map[string]any{}
+		paramHolder.append(params)
+		logCtx.WithField("cluster", cluster.Name).Debug("matched cluster secret")
+	}
 
-			if len(cluster.Annotations) > 0 {
-				meta["annotations"] = cluster.Annotations
-			}
-			if len(cluster.Labels) > 0 {
-				meta["labels"] = cluster.Labels
-			}
+	return paramHolder.consolidate(), nil
+}
 
-			params["metadata"] = meta
-		} else {
-			for key, value := range cluster.Annotations {
-				params["metadata.annotations."+key] = value
-			}
+type paramHolder struct {
+	isFlatMode bool
+	params     []map[string]any
+}
 
-			for key, value := range cluster.Labels {
-				params["metadata.labels."+key] = value
-			}
-		}
+func newParamHolder(isFlatMode bool) *paramHolder {
+	return &paramHolder{isFlatMode: isFlatMode}
+}
```

Comment:
Same for this function, given that this function is as a constructor and we only have one parameter, do we need to move it to a separate function?

Reply:
Good point, we can live without this.